matches/dao/match-dao.go

```diff
@@ -0,0 +1,172 @@
+package dao
+
+import (
+	// pq acts as the driver for SQL requests
+	"database/sql"
+	"errors"
+
+	_ "github.com/lib/pq"
+)
+
+const connStr = "user=postgres dbname=postgres host=matches-db sslmode=disable"
+
+// MatchGetResponse contains the information stored about a given match
+type MatchGetResponse struct {
+	ID        int
+	userOne   int
+	userTwo   int
+	matchedOn string
+}
+
+// MatchListResponse contains the information stored about all matches
+type MatchListResponse struct {
+	IDs []int
+}
+
+// MatchCreateRequest contains the information required to create a new match
+type MatchCreateRequest struct {
+	ID1 string `valid:"type(string),required,stringlength(1|255)"`
+	ID2 string `valid:"type(string),required,stringlength(1|255)"`
+}
+
+// MatchCreateResponse contains the information stored about the newly created match
+type MatchCreateResponse struct {
+	ID        int
+	userOne   int
+	userTwo   int
+	matchedOn string
+}
+
+type MatchUpdateRequest struct {
+	ID      int
+	userOne int
+	userTwo int
+}
+
+func executeQueryWithResponses(query string, args ...interface{}) (*sql.Rows, error) {
+	db, err := sql.Open("postgres", connStr)
+	if err != nil {
+		return nil, err
+	}
+	defer db.Close()
+	rows, err := db.Query(query, args...)
+	return rows, err
+}
+
+func executeQueryWithRowResponse(query string, args ...interface{}) (*sql.Row, error) {
+	db, err := sql.Open("postgres", connStr)
+	if err != nil {
+		return nil, err
+	}
+	defer db.Close()
+	return db.QueryRow(query, args...), nil
+}
+
+func executeQuery(query string, args ...interface{}) (int64, error) {
+	db, err := sql.Open("postgres", connStr)
+	if err != nil {
+		return 0, err
+	}
+	defer db.Close()
+
+	result, err := db.Exec(query, args...)
+	if err != nil {
+		return 0, err
+	}
+
+	return result.RowsAffected()
+}
+
+// GetMatch returns the information about a match stored for a given ID
+func GetMatch(id int64) (*MatchGetResponse, error) {
+	row, err := executeQueryWithRowResponse("SELECT * FROM Matches WHERE id = $1", id)
+	if err != nil {
+		return nil, err
+	}
+
+	var match MatchGetResponse
+	err = row.Scan(&match.ID, &match.userOne, &match.userTwo, &match.matchedOn)
+	if err != nil {
+		switch err {
+		case sql.ErrNoRows:
+			return nil, ErrMatchNotFound(id)
+		default:
+			return nil, err
+		}
+	}
+
+	return &match, nil
+}
+
+// ListMatch lists all the matches which feature a user
+func ListMatch(id1 int64) (*MatchListResponse, error) {
+	rows, err := executeQueryWithResponses("SELECT id FROM Matches WHERE userOne = $1 OR userTwo = $1", id1)
+	if err != nil {
+		return nil, err
+	}
+
+	var matches MatchListResponse
+	matches.IDs = make([]int, 0)
```

Comment:
Can we make a list of the correct size using the rows object? Not sure if it returns the number of rows found